Consider options for setting image creation time while maintaining reproducibility

Followup from discussion: https://github.com/GoogleContainerTools/jib/pull/412#discussion_r196172710

It doesn't work, when I config just like this : 
            <plugin>
                <groupId>com.google.cloud.tools</groupId>
                <artifactId>jib-maven-plugin</artifactId>
                <version>0.9.6</version>
                <configuration>
                    <to>
                        <image>myimage</image>
                    </to>

                   <container>
                       <useCurrentTimestamp>true</useCurrentTimestamp>
                   </container>
                </configuration>
            </plugin>


the error log : 
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:0.9.6:dockerBuild (default-cli) on project jib-demo: Unable to parse configuration of mojo com.google.cloud.tools:jib-maven-plugin:0.9.6:dockerBuild for parameter useCurrentTimestamp: Cannot find 'useCurrentTimestamp' in class com.google.cloud.tools.jib.maven.JibPluginConfiguration$ContainerParameters -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginConfigurationException

--- a/jib-maven-plugin/CHANGELOG.md
+++ b/jib-maven-plugin/CHANGELOG.md
@@ -7,13 +7,14 @@ All notable changes to this project will be documented in this file.
 
 - Snapshot dependencies are added as their own layer ([#584](https://github.com/GoogleContainerTools/jib/pull/584))
 - `jib:buildTar` goal to build an image tarball at `target/jib-image.tar`, which can be loaded into docker using `docker load` ([#514](https://github.com/GoogleContainerTools/jib/issues/514))
-- For Docker Hub, also tries registry aliases when getting a credential from the Docker config
+- `<container><useCurrentTimestamp>` parameter to set the image creation time to the build time ([#413](https://github.com/GoogleContainerTools/jib/issues/413))
 
 ### Changed
 
 ### Fixed
 
 - Using multi-byte characters in container configuration ([#626](https://github.com/GoogleContainerTools/jib/issues/626))
+- For Docker Hub, also tries registry aliases when getting a credential from the Docker config ([#605](https://github.com/GoogleContainerTools/jib/pull/605))
 
 ## 0.9.6
 

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildDockerMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildDockerMojo.java
@@ -33,6 +33,7 @@ import com.google.common.base.Preconditions;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.time.Instant;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import org.apache.maven.plugin.MojoExecutionException;
@@ -112,6 +113,12 @@ public class BuildDockerMojo extends JibPluginConfiguration {
       buildConfigurationBuilder.setBaseImageLayersCacheConfiguration(
           applicationLayersCacheConfiguration);
     }
+    if (getUseCurrentTimestamp()) {
+      mavenBuildLogger.warn(
+          "Setting image creation time to current time; your image may not be reproducible.");
+      buildConfigurationBuilder.setCreationTime(Instant.now());
+    }
+
     BuildConfiguration buildConfiguration = buildConfigurationBuilder.build();
 
     // TODO: Instead of disabling logging, have authentication credentials be provided

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildImageMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildImageMojo.java
@@ -34,6 +34,7 @@ import com.google.common.base.Strings;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.time.Instant;
 import java.util.Arrays;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
@@ -134,6 +135,12 @@ public class BuildImageMojo extends JibPluginConfiguration {
       buildConfigurationBuilder.setBaseImageLayersCacheConfiguration(
           applicationLayersCacheConfiguration);
     }
+    if (getUseCurrentTimestamp()) {
+      mavenBuildLogger.warn(
+          "Setting image creation time to current time; your image may not be reproducible.");
+      buildConfigurationBuilder.setCreationTime(Instant.now());
+    }
+
     BuildConfiguration buildConfiguration = buildConfigurationBuilder.build();
 
     // TODO: Instead of disabling logging, have authentication credentials be provided

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildTarMojo.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildTarMojo.java
@@ -33,6 +33,7 @@ import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.time.Instant;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import org.apache.maven.plugin.MojoExecutionException;
@@ -110,6 +111,12 @@ public class BuildTarMojo extends JibPluginConfiguration {
       buildConfigurationBuilder.setBaseImageLayersCacheConfiguration(
           applicationLayersCacheConfiguration);
     }
+    if (getUseCurrentTimestamp()) {
+      mavenBuildLogger.warn(
+          "Setting image creation time to current time; your image may not be reproducible.");
+      buildConfigurationBuilder.setCreationTime(Instant.now());
+    }
+
     BuildConfiguration buildConfiguration = buildConfigurationBuilder.build();
 
     // TODO: Instead of disabling logging, have authentication credentials be provided

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java
@@ -67,6 +67,8 @@ abstract class JibPluginConfiguration extends AbstractMojo {
   /** Configuration for {@code container} parameter. */
   public static class ContainerParameters {
 
+    @Parameter private boolean useCurrentTimestamp = false;
+
     @Parameter private List<String> jvmFlags = Collections.emptyList();
 
     @Nullable @Parameter private String mainClass;
@@ -189,6 +191,10 @@ abstract class JibPluginConfiguration extends AbstractMojo {
   @Nullable
   String getTargetImageCredentialHelperName() {
     return Preconditions.checkNotNull(to).credHelper;
+  }
+
+  boolean getUseCurrentTimestamp() {
+    return container.useCurrentTimestamp;
   }
 
   List<String> getJvmFlags() {

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildDockerMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildDockerMojoIntegrationTest.java
@@ -19,6 +19,7 @@ package com.google.cloud.tools.jib.maven;
 import com.google.cloud.tools.jib.Command;
 import java.io.IOException;
 import java.nio.file.Path;
+import java.time.Instant;
 import org.apache.maven.it.VerificationException;
 import org.apache.maven.it.Verifier;
 import org.hamcrest.CoreMatchers;
@@ -68,11 +69,23 @@ public class BuildDockerMojoIntegrationTest {
 
   @Test
   public void testExecute_simple() throws VerificationException, IOException, InterruptedException {
+    Instant before = Instant.now();
     Assert.assertEquals(
         "Hello, world. An argument.\nfoo\ncat\n",
         buildToDockerDaemonAndRun(
             simpleTestProject.getProjectRoot(),
             "gcr.io/jib-integration-testing/simpleimage:maven"));
+    Instant buildTime =
+        Instant.parse(
+            new Command(
+                    "docker",
+                    "inspect",
+                    "-f",
+                    "{{.Created}}",
+                    "gcr.io/jib-integration-testing/simpleimage:maven")
+                .run()
+                .trim());
+    Assert.assertTrue(buildTime.isAfter(before) || buildTime.equals(before));
   }
 
   @Test
@@ -81,6 +94,16 @@ public class BuildDockerMojoIntegrationTest {
         "",
         buildToDockerDaemonAndRun(
             emptyTestProject.getProjectRoot(), "gcr.io/jib-integration-testing/emptyimage:maven"));
+    Assert.assertEquals(
+        "1970-01-01T00:00:00Z",
+        new Command(
+                "docker",
+                "inspect",
+                "-f",
+                "{{.Created}}",
+                "gcr.io/jib-integration-testing/emptyimage:maven")
+            .run()
+            .trim());
   }
 
   @Test

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java
@@ -19,6 +19,7 @@ package com.google.cloud.tools.jib.maven;
 import com.google.cloud.tools.jib.Command;
 import java.io.IOException;
 import java.nio.file.Path;
+import java.time.Instant;
 import java.util.Arrays;
 import org.apache.maven.it.VerificationException;
 import org.apache.maven.it.Verifier;
@@ -101,11 +102,24 @@ public class BuildImageMojoIntegrationTest {
                   + "clean jib:build\" instead of \"mvn clean compile jib:build\"?)"));
     }
 
+    Instant before = Instant.now();
     Assert.assertEquals(
         "Hello, world. An argument.\nfoo\ncat\n",
         buildAndRun(
             simpleTestProject.getProjectRoot(),
             "gcr.io/jib-integration-testing/simpleimage:maven"));
+
+    Instant buildTime =
+        Instant.parse(
+            new Command(
+                    "docker",
+                    "inspect",
+                    "-f",
+                    "{{.Created}}",
+                    "gcr.io/jib-integration-testing/simpleimage:maven")
+                .run()
+                .trim());
+    Assert.assertTrue(buildTime.isAfter(before) || buildTime.equals(before));
   }
 
   @Test
@@ -114,6 +128,16 @@ public class BuildImageMojoIntegrationTest {
         "",
         buildAndRun(
             emptyTestProject.getProjectRoot(), "gcr.io/jib-integration-testing/emptyimage:maven"));
+    Assert.assertEquals(
+        "1970-01-01T00:00:00Z",
+        new Command(
+                "docker",
+                "inspect",
+                "-f",
+                "{{.Created}}",
+                "gcr.io/jib-integration-testing/emptyimage:maven")
+            .run()
+            .trim());
   }
 
   @Test

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildTarMojoIntegrationTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/BuildTarMojoIntegrationTest.java
@@ -18,6 +18,7 @@ package com.google.cloud.tools.jib.maven;
 
 import com.google.cloud.tools.jib.Command;
 import java.io.IOException;
+import java.time.Instant;
 import org.apache.maven.it.VerificationException;
 import org.apache.maven.it.Verifier;
 import org.junit.Assert;
@@ -37,6 +38,7 @@ public class BuildTarMojoIntegrationTest {
    */
   @Test
   public void testExecute_simple() throws VerificationException, IOException, InterruptedException {
+    Instant before = Instant.now();
     Verifier verifier = new Verifier(simpleTestProject.getProjectRoot().toString());
     verifier.setAutoclean(false);
     verifier.executeGoal("package");
@@ -57,5 +59,17 @@ public class BuildTarMojoIntegrationTest {
     Assert.assertEquals(
         "Hello, world. An argument.\nfoo\ncat\n",
         new Command("docker", "run", "gcr.io/jib-integration-testing/simpleimage:maven").run());
+
+    Instant buildTime =
+        Instant.parse(
+            new Command(
+                    "docker",
+                    "inspect",
+                    "-f",
+                    "{{.Created}}",
+                    "gcr.io/jib-integration-testing/simpleimage:maven")
+                .run()
+                .trim());
+    Assert.assertTrue(buildTime.isAfter(before) || buildTime.equals(before));
   }
 }

--- a/jib-maven-plugin/src/test/resources/projects/simple/pom.xml
+++ b/jib-maven-plugin/src/test/resources/projects/simple/pom.xml
@@ -42,6 +42,7 @@
             <image>gcr.io/jib-integration-testing/simpleimage:maven</image>
           </to>
           <container>
+            <useCurrentTimestamp>true</useCurrentTimestamp>
             <args>An argument.</args>
             <ports>
               <port>1000/tcp</port>

